library/ssl_cli.c

```diff
@@ -671,6 +671,126 @@ static void ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
+                        MBEDTLS_SRTP_AES128_CM_HMAC_SHA1_80_IANA_VALUE ) );
+                *p++ = ( ( ( MBEDTLS_SRTP_AES128_CM_HMAC_SHA1_80_IANA_VALUE )
+                            >> 8 ) & 0xFF );
+                *p++ = ( ( MBEDTLS_SRTP_AES128_CM_HMAC_SHA1_80_IANA_VALUE )
+                          & 0xFF );
+                break;
+            case MBEDTLS_SRTP_AES128_CM_HMAC_SHA1_32:
+                MBEDTLS_SSL_DEBUG_MSG( 3, ( "ssl_write_use_srtp_ext, add profile: %04x",
+                        MBEDTLS_SRTP_AES128_CM_HMAC_SHA1_32_IANA_VALUE ) );
+                *p++ = ( ( ( MBEDTLS_SRTP_AES128_CM_HMAC_SHA1_32_IANA_VALUE )
+                            >> 8 ) & 0xFF );
+                *p++ = ( ( MBEDTLS_SRTP_AES128_CM_HMAC_SHA1_32_IANA_VALUE )
+                         & 0xFF );
+                break;
+            case MBEDTLS_SRTP_NULL_HMAC_SHA1_80:
+                MBEDTLS_SSL_DEBUG_MSG( 3, ( "ssl_write_use_srtp_ext, add profile: %04x",
+                        MBEDTLS_SRTP_NULL_HMAC_SHA1_80_IANA_VALUE ) );
+                *p++ = ( ( ( MBEDTLS_SRTP_NULL_HMAC_SHA1_80_IANA_VALUE ) >> 8 )
+                         & 0xFF ) ;
+                *p++ = ( ( MBEDTLS_SRTP_NULL_HMAC_SHA1_80_IANA_VALUE ) & 0xFF );
+                break;
+            case MBEDTLS_SRTP_NULL_HMAC_SHA1_32:
+                MBEDTLS_SSL_DEBUG_MSG( 3, ( "ssl_write_use_srtp_ext, add profile: %04x",
+                        MBEDTLS_SRTP_NULL_HMAC_SHA1_32_IANA_VALUE ) );
+                *p++ = ( ( ( MBEDTLS_SRTP_NULL_HMAC_SHA1_32_IANA_VALUE ) >> 8 )
+                         & 0xFF );
+                *p++ = ( ( MBEDTLS_SRTP_NULL_HMAC_SHA1_32_IANA_VALUE ) & 0xFF );
+                break;
+            default:
+                /*
+                 * Note: we shall never arrive here as protection profiles
+                 * is checked by ssl_set_dtls_srtp_protection_profiles function
+                 */
+                MBEDTLS_SSL_DEBUG_MSG( 1, ( "client hello, ignore illegal DTLS-SRTP protection profile %d",
+                                            ssl->conf->dtls_srtp_profile_list[protection_profiles_index]) );
+                break;
+        }
+    }
+
+    *p++ = mki_len & 0xFF;
+
+    if( mki_len != 0 )
+    {
+        for( i=0; i < mki_len; i++ )
+        {
+            *p++ = ssl->dtls_srtp_info.mki_value[i];
+        }
+        MBEDTLS_SSL_DEBUG_BUF( 3, "sending mki",  ssl->dtls_srtp_info.mki_value,
+                               ssl->dtls_srtp_info.mki_len );
+    }
+
+    /*
+     * total extension length: extension type (2 bytes)
+     *                         + extension length (2 bytes)
+     *                         + protection profile length (2 bytes)
+     *                         + 2 * number of protection profiles
+     *                         + srtp_mki vector length(1 byte)
+     */
+    *olen = 2 + 2 + 2 + 2 * ( ssl->conf->dtls_srtp_profile_list_len ) + 1 + mki_len;
```

Comment:
Since the length field is a t the beginning of the message, your suggestion will require either:
1. having a temporary buffer to hold all the data
2.  having `p` point to `buf + 4` and then writing the length to `buf[2]` and `buf[3]`

I am not sure this is a better approach.
I can accept though changing this line to improve readability.( either `p - buf` or `ext_len + 4`)

Reply:
@RonEld That's not what I meant. I mean not hardcoding the total length here - you could either add a local variable `total_len` which is incremented as you write the extension, and write that into `*olen` in the end, or alternatively use something like `p - buf` to see how much you have written.
Note that this is different from writing the length values in the target buffer; if you want to do that, you need to - as you say - remember the corresponding part of the buffer until the end. I think that's sensible, too, but requires more work and it not part of the request this comment makes.

Reply:
ok. My initial understanding was to write to the buffer only once everything is validated, but as you said, it's not part of this scope.
I changed it to `p - buf`